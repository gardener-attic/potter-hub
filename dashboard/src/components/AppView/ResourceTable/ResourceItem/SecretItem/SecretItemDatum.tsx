import * as React from "react";
import { Copy, Eye, EyeOff } from "react-feather";

interface ISecretItemDatumProps {
  name: string;
  value: string;
}

interface ISecretItemDatumState {
  hidden: boolean;
}

class SecretItemDatum extends React.PureComponent<ISecretItemDatumProps, ISecretItemDatumState> {
  // Secret datum is hidden by default
  public state: ISecretItemDatumState = {
    hidden: true,
  };

  public render() {
    const { name, value } = this.props;
    const { hidden } = this.state;
    const decodedValue = atob(value);
    return (
      <span className="flex">
        <a id="togglesecret" onClick={this.toggleDisplay} title="Toggle secret">
          {hidden ? <Eye /> : <EyeOff />}
        </a>
        <span className="flex margin-l-normal">
          <a id="copysecret" onClick={this.copySecret} title="Copy plain secret to clipboard">
            <Copy />
          </a>
        </span>
        <span className="flex margin-l-normal">
          <span>{name}:</span>
          {hidden ? (
            <span className="margin-l-small">{decodedValue.length} bytes</span>
          ) : (
            <pre className="SecretContainer">
              <code className="SecretContent">{this.shortenSecret()}</code>
            </pre>
          )}
        </span>
      </span>
    );
  }

  private copySecret = () => {
    const value = this.props.value;
    const decodedValue = atob(value);
    navigator.clipboard.writeText(decodedValue);
  };

  private shortenSecret = () => {
    const maxLength = 128;
    const value = this.props.value;
    const decodedValue = atob(value);
    if (decodedValue.length > maxLength) {
      return decodedValue.substring(0, maxLength - 3) + "...";
    }
    return decodedValue;
  };

  private toggleDisplay = () => {
    this.setState({
      hidden: !this.state.hidden,
    });
  };
}

export default SecretItemDatum;
