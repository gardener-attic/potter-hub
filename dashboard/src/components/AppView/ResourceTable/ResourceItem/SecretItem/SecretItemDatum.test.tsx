import { shallow } from "enzyme";
import * as React from "react";
import SecretItemDatum from "./SecretItemDatum";

const testProps = {
  name: "foo",
  value: "YmFy", // foo
};

// First 125 characters
const shortenedSecret =
  "Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna ali";
const testLongProps = {
  name: "foo",
  // The original secret is 300 characters long
  value:
    "TG9yZW0gaXBzdW0gZG9sb3Igc2l0IGFtZXQsIGNvbnNldGV0dXIgc2FkaXBzY2luZyBlbGl0ciwgc2VkIGRpYW0gbm9udW15IGVpcm1vZCB0ZW1wb3IgaW52aWR1bnQgdXQgbGFib3JlIGV0IGRvbG9yZSBtYWduYSBhbGlxdXlhbSBlcmF0LCBzZWQgZGlhbSB2b2x1cHR1YS4gQXQgdmVybyBlb3MgZXQgYWNjdXNhbSBldCBqdXN0byBkdW8gZG9sb3JlcyBldCBlYSByZWJ1bS4gU3RldCBjbGl0YSBrYXNkIGd1YmVyZ3Jlbiwgbm8gc2VhIHRha2ltYXRhIHNhbmN0dXMgZXN0IExvcmVtIGlwc3VtIGRvbG9yIHNpdCBhbWV0LiBMb3Jl",
};

it("renders the secret datum (hidden by default)", () => {
  const wrapper = shallow(<SecretItemDatum {...testProps} />);
  expect(wrapper.state()).toMatchObject({ hidden: true });
  expect(wrapper).toMatchSnapshot();
});

it("displays the secret datum value when clicking on the icon", () => {
  const wrapper = shallow(<SecretItemDatum {...testProps} />);
  expect(wrapper.text()).toContain("foo:3 bytes");
  const icon = wrapper.find("a#togglesecret");
  expect(icon).toExist();
  icon.simulate("click");
  expect(wrapper.state()).toMatchObject({ hidden: false });
  expect(wrapper.text()).toContain("foo:bar");
});

it("displays the secret datum shortened when clicking on the icon", () => {
  const wrapper = shallow(<SecretItemDatum {...testLongProps} />);
  expect(wrapper.text()).toContain("foo:300 bytes");
  const icon = wrapper.find("a#togglesecret");
  expect(icon).toExist();
  icon.simulate("click");
  expect(wrapper.state()).toMatchObject({ hidden: false });
  expect(wrapper.text()).toContain("foo:" + shortenedSecret);
});
