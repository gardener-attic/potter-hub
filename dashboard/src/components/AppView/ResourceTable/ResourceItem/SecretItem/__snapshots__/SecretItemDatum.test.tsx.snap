// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the secret datum (hidden by default) 1`] = `
<span
  className="flex"
>
  <a
    id="togglesecret"
    onClick={[Function]}
    title="Toggle secret"
  >
    <Eye
      color="currentColor"
      size="24"
    />
  </a>
  <span
    className="flex margin-l-normal"
  >
    <a
      id="copysecret"
      onClick={[Function]}
      title="Copy plain secret to clipboard"
    >
      <Copy
        color="currentColor"
        size="24"
      />
    </a>
  </span>
  <span
    className="flex margin-l-normal"
  >
    <span>
      foo
      :
    </span>
    <span
      className="margin-l-small"
    >
      3
       bytes
    </span>
  </span>
</span>
`;
